plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesbuilder/N4JSClassDeclarationTypesBuilder.xtend

```diff
@@ -10,32 +10,33 @@
  */
 package org.eclipse.n4js.typesbuilder
 
-import com.google.inject.Inject
 import org.eclipse.n4js.AnnotationDefinition
 import org.eclipse.n4js.n4JS.N4ClassDeclaration
 import org.eclipse.n4js.n4JS.N4ClassDefinition
 import org.eclipse.n4js.n4JS.N4ClassExpression
-import org.eclipse.n4js.n4JS.N4FieldDeclaration
-import org.eclipse.n4js.n4JS.N4GetterDeclaration
-import org.eclipse.n4js.n4JS.N4MethodDeclaration
-import org.eclipse.n4js.n4JS.N4SetterDeclaration
-import org.eclipse.n4js.ts.typeRefs.ParameterizedTypeRef
 import org.eclipse.n4js.ts.types.TClass
 import org.eclipse.n4js.ts.types.TModule
-import org.eclipse.n4js.ts.types.TypeAccessModifier
 import org.eclipse.n4js.ts.types.TypesFactory
 import org.eclipse.n4js.ts.types.TypingStrategy
-import java.util.List
+import org.eclipse.n4js.ts.utils.TypeUtils
 
 import static extension org.eclipse.n4js.utils.N4JSLanguageUtils.*
 
-public class N4JSClassDeclarationTypesBuilder {
+public class N4JSClassDeclarationTypesBuilder extends N4JSClassifierDeclarationTypesBuilder {
 
-	@Inject extension N4JSTypesBuilderHelper
-	@Inject extension N4JSFieldTypesBuilder
-	@Inject extension N4JSMethodTypesBuilder
-	@Inject extension N4JSGetterTypesBuilder
-	@Inject extension N4JSSetterTypesBuilder
+	def protected boolean linkTClass(N4ClassDeclaration n4Class, TModule target, boolean preLinkingPhase, int idx) {
+		if (n4Class.name === null) { // may be null due to syntax errors
+			return false;
+		}
+
+		val TClass tclass = target.topLevelTypes.get(idx) as TClass
+		if (tclass.name != n4Class.name) {
+			throw new IllegalStateException('name mismatch: ' + tclass.name + ' vs ' + n4Class.name);
```

Comment:
here you check the name; but in many other cases no such check is performed (e.g. methods, fpars, exported variables)

What was the general approach?
1. make `N4JSTypesBuilder#linkTypes()` resilient and recognize mismatching old/new AST in types builder (and thus *not* requiring an md5 hash)
2. use an md5 hash and then just assuming in `N4JSTypesBuilder#linkTypes()` that there isn't a mismatch (in that case the above name check is unnecessary)

Reply:
We should always check the name for equality. That should do the trick. For lists, we could consider to also check the length before we traverse them.